aarch64: build fails with TestOverlay128LayerRead

**Description**

I'm trying to build master on ARMv8 (aarch64). The overlay2 test fails with "TestOverlay128LayerRead".

**Steps to reproduce the issue:**
1. install Docker on Ubuntu for aarch64 with "apt-get install docker.io"
2. download docker source and build with "make all"

**Describe the results you received:**

```
-- FAIL: TestOverlay128LayerRead (4.71s)
FAIL
FAIL    github.com/docker/docker/daemon/graphdriver/overlay2    5.506s
```

**Describe the results you expected:**

Expect all tests to pass.

**Additional information you deem important (e.g. issue happens only occasionally):**

Repeatable on successive builds.

**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   23cf638
 Built:        Tue, 27 Sep 2016 12:25:38 +1300
 OS/Arch:      linux/arm64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   23cf638
 Built:        Tue, 27 Sep 2016 12:25:38 +1300
 OS/Arch:      linux/arm64
```

**Output of `docker info`:**

```
Containers: 7
 Running: 0
 Paused: 0
 Stopped: 7
Images: 117
Server Version: 1.12.1
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge overlay null host
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor
Kernel Version: 4.4.0-38-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: aarch64
CPUs: 96
Total Memory: 125.8 GiB
Name: armv8hello.local.lan
ID: QB72:5PAI:O62E:QEPU:4DKN:3M2K:33FC:AIJU:PXMC:YMGX:Y376:IAPK
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Hosted ARMv8 96 core server

The dsdbench tool also has a benchmark tool, whose output is at

https://gist.github.com/e7dafbd07faf4a20a8f829254810de2d


```
BenchmarkGet100BaseMount-96     --- FAIL: BenchmarkGet100BaseMount-96
        bench_test.go:110: Failed to create layer chain: error creating overlay mount to /tmp/layer-test-149274786/overlay2/5e2c43486c8f17fd3b62738ed78d14a27bee1cd9392d8e89ea4c89e4709b4bea/merged: no such file or directory
                failed to mount
                dsdbench.CreateLayer
                        /root/src/dsdbench/test_util.go:38
                dsdbench.CreateLayerChain
                        /root/src/dsdbench/test_util.go:72
                dsdbench.benchmarkGetBaseMountWithDepth
                        /root/src/dsdbench/bench_test.go:108
                dsdbench.BenchmarkGet100BaseMount
                        /root/src/dsdbench/bench_test.go:84
                testing.(*B).runN
                        /usr/lib/go-1.6/src/testing/benchmark.go:135
                testing.(*B).launch
                        /usr/lib/go-1.6/src/testing/benchmark.go:210
                runtime.goexit
                        /usr/lib/go-1.6/src/runtime/asm_arm64.s:975
                failed to create layer 62
                dsdbench.CreateLayerChain
                        /root/src/dsdbench/test_util.go:75
                dsdbench.benchmarkGetBaseMountWithDepth
                        /root/src/dsdbench/bench_test.go:108
                dsdbench.BenchmarkGet100BaseMount
                        /root/src/dsdbench/bench_test.go:84
                testing.(*B).runN
                        /usr/lib/go-1.6/src/testing/benchmark.go:135
                testing.(*B).launch
                        /usr/lib/go-1.6/src/testing/benchmark.go:210
                runtime.goexit
```

--- a/daemon/graphdriver/overlay2/overlay.go
+++ b/daemon/graphdriver/overlay2/overlay.go
@@ -473,6 +473,15 @@ func (d *Driver) Get(id string, mountLabel string) (s string, err error) {
 
 	pageSize := syscall.Getpagesize()
 
+	// Go can return a larger page size than supported by the system
+	// as of go 1.7. This will be fixed in 1.8 and this block can be
+	// removed when building with 1.8.
+	// See https://github.com/golang/go/commit/1b9499b06989d2831e5b156161d6c07642926ee1
+	// See https://github.com/docker/docker/issues/27384
+	if pageSize > 4096 {
+		pageSize = 4096
+	}
+
 	// Use relative paths and mountFrom when the mount data has exceeded
 	// the page size. The mount syscall fails if the mount data cannot
 	// fit within a page and relative links make the mount data much

